common-tools/clas-detector/src/main/java/org/jlab/detector/decode/CodaEventDecoder.java

```diff
@@ -1049,34 +1049,60 @@ public HelicityDecoderData getDataEntries_HelicityDecoder(EvioDataEvent event){
                     int[]  intData  = ByteDataTransformer.toIntArray(node.getStructureBuffer(true));
-                    long  timeStamp = longData[2]&0x0000ffffffffffffL;
-                    
-                    int tsettle  = DataUtils.getInteger(intData[16], 0, 0) > 0 ? 1 : -1;
-                    int pattern  = DataUtils.getInteger(intData[16], 1, 1) > 0 ? 1 : -1;
-                    int pair     = DataUtils.getInteger(intData[16], 2, 2) > 0 ? 1 : -1;
-                    int helicity = DataUtils.getInteger(intData[16], 3, 3) > 0 ? 1 : -1;                             
-                    int start    = DataUtils.getInteger(intData[16], 4, 4) > 0 ? 1 : -1;                             
-                    int polarity = DataUtils.getInteger(intData[16], 5, 5) > 0 ? 1 : -1;                             
-                    int count    = DataUtils.getInteger(intData[16], 8, 11);
+
+                    // When there are multiple HelicityDecoder banks in an event, there is a BLKHDR work in the data,
+                    // and when there is one HelicityDecoder bank in an event, it is not there. So we need to
+                    // detect where the trigger time word is.
+                    int i_data_offset = 2;
+                    int i_data_length = intData.length;
+                    while(i_data_offset<intData.length){
+                        // The following idiotic construction is needed because Java doesn't have unsigned ints,
+                        // and a right shift on a negative int results in a negative number.
+                        int int_test_value = (int) (( ((long)intData[i_data_offset]) & 0x00000000ffffffffL ) >> 27);
+                        if(int_test_value == 0x13) break;
+                        i_data_offset++;
+                    } // find the trigger time word.
+                    if(i_data_offset>=intData.length){
+                        System.err.println("ERROR:  HelicityDecoder data is corrupted. Trigger time word not found.");
+                        return null;
+                    }
+                    long  timeStamp = (intData[i_data_offset]&0x00ffffff) + (((long)(intData[i_data_offset+1]&0x00ffffffL))<<24);
+                    i_data_offset+=2; // Next word should be "DECODER DATA", with 0x18 in the top 5 bits.
+                    if(((int) (( ((long)intData[i_data_offset]) & 0x00000000ffffffffL ) >> 27)) != 0x18){
+                        System.err.println("ERROR:  HelicityDecoder data is corrupted. DECODER BANK not found.");
+                        return null;
+                    }
+                    int num_data_words = intData[i_data_offset]&0x07ffffff;
+                    if(num_data_words < 14){
+                        System.err.println("ERROR:  HelicityDecoder data is corrupted. Not enough data words.");
+                        return null;
+                    }
+                    int tsettle  = DataUtils.getInteger(intData[i_data_offset+9], 0, 0) > 0 ? 1 : -1;
```

Comment:
i_data_offset should be incremented by 1 before this line

Reply:
Thank you for catching that Raffaella. You are quite correct. An update was pushed.
I still need to do a detailed analysis of the helicity. I am not quite understanding what Nathan is referring to.

Reply:
I'm referring to the most critical data in the HEL::decoder bank, those "patternArray" variables that contain the helicity state sequences seen so far.  Previously they were consistent with good sequences.